rllib/core/optim/marl_optimizer.py

```diff
@@ -0,0 +1,83 @@
+from typing import Any, Dict, Mapping, Union, Type
+from ray.rllib.core.rl_module.marl_module import MultiAgentRLModule
+from ray.rllib.core.rl_module.rl_module import ModuleID
+from ray.rllib.core.optim.rl_optimizer import RLOptimizer
+from ray.rllib.policy.sample_batch import DEFAULT_POLICY_ID
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.typing import TensorType
+
+
+class MultiAgentRLOptimizer(RLOptimizer):
+    def __init__(self, rl_optimizers: Mapping[ModuleID, RLOptimizer] = None):
+        self._optimizers = rl_optimizers
+
+    @classmethod
+    def from_config(
```

Comment:
we can totally do this, but how do we make it so that people can use these new custom entry points with the RLTrainer class.

Reply:
You gotta think about these in the top down context, i.e. when you are instantiating them what are the possible scenarios and in each scenario how would we instantiate the object without restricting the user to too many constraints on the design of sub-classes. Can we consolidate these scenarios into one generic APIs to make the cognitive load minimal? 

My thinking process was that RLTrainer will use one or the other, we can devise new entry points in new contexts when the time comes. It will give us more flexibility for later. 
For example, for now we will always use MARLOptimizer.from_marl_module() in RLTrainer if the rl_optimizer_class is not a MARLOptimizer already (hence it's SARLOptimizer). We will use the other entry point if it's a single RL Optimizer and then call as_multi_agent() on it? I haven't worked out the entire stack tho. I was hoping this would give you starting ideas on how to modify these entry points. But my point is that the entry point should support both SARL and MARL and different possible scenarios before we can merge this PR. That's the ultimate goal. It should also be generic enough so that people can use XXXOptimizer in multi agent and single agent context in their own none-RLlib code.

Reply:
There are 3 things we can try do here:

1) Make identical entry points for both optimizers,This is reasonable if they both just taking in a module and a config 
2) Make an opinionated choice to only allow the RLTrainer to accept `MARLModule` and `MARLOptim`.
      - Add something to the `RLTrainer` class that converts the constructor of the class to the constructor of a MARLOptim class

Reply:
As long as we are consistent 2 makes the most sense with the idea of MARL should be first class citizen. If SARL is the actual use-case the caller has to convert it to MultiAgent first and pass it in?

Reply:
I was able to make things work for now with different entry points. 
Although its still unclear how users will specify their entry point. maybe we can give them a function that they can pass for doing construction.